src/dimension.c

```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
+					 errmsg("invalid number of partitions: must be between 1 and %d", INT16_MAX)));
+	}
+	else
+	{
+		/* Open ("time") dimension */
+		info->type = DIMENSION_TYPE_OPEN;
+		info->set_not_null = !not_null_is_set;
+
+		if (!IS_VALID_OPEN_DIM_TYPE(info->coltype))
+			ereport(ERROR,
+					(errcode(ERRCODE_WRONG_OBJECT_TYPE),
+					 errmsg("invalid dimension type: must be an integer, date or timestamp")));
+
+		if (!OidIsValid(info->interval_type))
+		{
+			if (IS_INTEGER_TYPE(info->coltype))
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("integer dimensions require an explicit interval")));
+
+			info->interval_datum = Int64GetDatum(DEFAULT_CHUNK_TIME_INTERVAL);
+			info->interval_type = INT8OID;
+		}
+
+		info->interval = dimension_interval_to_internal(info->coltype,
+														info->interval_type,
+														info->interval_datum,
+														info->show_warnings);
+
+		if (info->coltype == DATEOID &&
+			(info->interval <= 0 || info->interval % USECS_PER_DAY != 0))
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid interval: must be multiples of one day")));
+	}
+}
+
+TS_FUNCTION_INFO_V1(dimension_validate);
+
+Datum
+dimension_validate(PG_FUNCTION_ARGS)
+{
+	DimensionInfo info = {
+		.table_relid = PG_GETARG_OID(0),
+		.colname = PG_ARGISNULL(1) ? NULL : PG_GETARG_NAME(1),
+		.num_slices = PG_ARGISNULL(2) ? DatumGetInt32(-1) : PG_GETARG_INT32(2),
+		.num_slices_is_set = !PG_ARGISNULL(2),
+		.interval_datum = PG_ARGISNULL(3) ? DatumGetInt32(-1) : PG_GETARG_DATUM(3),
+		.interval_type = PG_ARGISNULL(3) ? InvalidOid : get_fn_expr_argtype(fcinfo->flinfo, 3),
+		.partitioning_func = PG_ARGISNULL(4) ? InvalidOid : PG_GETARG_OID(4),
+	};
+
+	dimension_validate_info(&info);
+
+	PG_RETURN_VOID();
+}
+
+TS_FUNCTION_INFO_V1(dimension_add);
+
+/*
+ * Add a new dimension to a hypertable.
+ *
+ * Arguments:
+ * 0. Relation ID of table
+ * 1. Column name
+ * 2. Number of partitions / slices in close ('space') dimensions
+ * 3. Interval for open ('time') dimensions
+ * 4. Partitioning function
+ * 5. IF NOT EXISTS option (bool)
+ */
+Datum
+dimension_add(PG_FUNCTION_ARGS)
+{
+	Cache	   *hcache = hypertable_cache_pin();
+	DimensionInfo info = {
+		.table_relid = PG_GETARG_OID(0),
+		.colname = PG_ARGISNULL(1) ? NULL : PG_GETARG_NAME(1),
+		.num_slices = PG_ARGISNULL(2) ? DatumGetInt32(-1) : PG_GETARG_INT32(2),
+		.num_slices_is_set = !PG_ARGISNULL(2),
+		.interval_datum = PG_ARGISNULL(3) ? DatumGetInt32(-1) : PG_GETARG_DATUM(3),
+		.interval_type = PG_ARGISNULL(3) ? InvalidOid : get_fn_expr_argtype(fcinfo->flinfo, 3),
+		.partitioning_func = PG_ARGISNULL(4) ? InvalidOid : PG_GETARG_OID(4),
+		.if_not_exists = PG_ARGISNULL(5) ? false : PG_GETARG_BOOL(5),
+		.ht = hypertable_cache_get_entry(hcache, PG_GETARG_OID(0)),
+		.show_warnings = true,
+	};
+
+	hypertable_permissions_check(info.table_relid, GetUserId());
+
+	if (!hypertable_lock_tuple_simple(info.table_relid))
```

Comment:
I am not quite sure why this lock is necessary. Why doesn't this code simply use `hypertable_increment_num_dimensions` instead of `hypertable_set_num_dimensions` and then not have to deal with this locking?

Reply:
Yeah so that is one option I certainly considered. A couple of reasons:

1. I tried to follow the old `plpgsql` code that took a row-level lock on the hypertable tuple.
2. We have a `CHECK` constraint on the `hypertable` table that says `num_dimensions > 0`, so `create_hypertable` must set it to at least 1. Therefore, it is not possible to use `hypertable_increment_num_dimensions` since it is not possible to know when `num_dimensions` has already been incremented (e.g., when called from `create_hypertable`) and when not (e.g., when this function is called explicitly). (Still, not sure that the catalog update API enforces check constraints, but makes sense to follow). The old code used a boolean to distinguish between these two cases, which seemed a bit ugly and unnecessary (especially if exposed in a public user-facing API).
3. An explicit row-level lock here makes it easier to bail out with a proper error message. (Although I guess we can take a table-level `RowExclusiveLock` already here)
4. `hypertable_increment_num_dimensions` and `hypertable_set_num_dimensions` take a table-level `RowExclusiveLock` and if the update fails, a general locking error ("tuple already updated") will be thrown that is more obscure. We could capture this error in a `TRY()` block an display a more appropriate error message though.

Open for suggestions though...

Reply:
Ahh ok. The `num_dimensions > 0` check constraint makes this more clear and persuasive. Can we add in a comment?